feat: adds Terraform Controller blog post (#16)
masterpointio/masterpoint.io: content/updates/terraform-controller-overview.md

---
visible: true
draft: false
title: "Mastering GitOps with Terraform Controller"
author: Veronika Gnilitska
slug: terraform-controller-overview
date: TBD
description: The post explores how Weave's Terraform Controller can enable GitOps for infrastructure management.
image: /img/updates/terraform-controller-overview/tf_controller_0.png
callout: "<p>👋 <b>Interested in platform engineering for your organization</b>, but not sure where to start? <a href='/contact'>Get in touch,</a> we're an expert team of platform engineers who deliver high-quality cloud platforms for startups and SMBs looking to scale. We enable your application engineers to focus on your product and in turn generate more value for your business.</p><a href='/contact' class='button'>Get In Touch &rsaquo;</a>"
---

> <h2>Table of Contents</h2>

Reviewer: Doesn't seem like this TOC is rendering correctly in the preview... we may need to cut it unless you know how it should be working?

Author: I was missing an empty row there, this is fixed. However, these extra spaces in the least important nested headings drive me nuts. I've noticed the similar behavior in the text, so I guess it's something we can fix on netlify side (?).
<img width="678" alt="Screenshot 2024-01-30 at 16 04 20" src="https://github.com/masterpointio/masterpoint.io/assets/30597968/6a50f764-5143-4564-978f-a0a933b6b7d1">

I love TOCs, they really help to evaluate if I'm interested in the content. But I have doubts if we should release it like that.

Possible sources for your inspiration are:

* [Ways of structuring your repositories](https://fluxcd.io/flux/guides/repository-structure/) by Flux team.
* [TF-controller demo: GitOps EKS node group](https://github.com/tf-controller/eks-scaling) by the Controller developers.

Both examples rely on Kustomize, which we've found somewhat inflexible regarding manifest templating. However, this method could certainly be highly effective for other organizations.

## Managing Infrastructure-as-Code with Terraform Controller

At its core, the Terraform Controller consists of two main elements: the Controller itself and the Runners.

The **Controller** is the brain of the operation. It monitors Kubernetes for specific resource definitions and acts upon changes. When a new Terraform resource is defined in Kubernetes, the Controller picks it up and initiates the necessary actions.

**Runners**, on the other hand, are the workhorses. Each Runner is a "disposable" pod (see `*-tf-runner` pod on the screenshot below) that executes the Terraform plans generated by the Controller.

![Runner of our ECR managed by Terraform Controller](/img/updates/terraform-controller-overview/tf_controller_3.png)

> The example of the Runner logs:

Reviewer: @gberenice this image is pretty small type in the website preview. Maybe we just include a few lines of the logs in a code block? What do you think?

Author: Updated. Let me know if that looks better to you.

Understanding these components and their roles is important for effectively leveraging the Terraform Controller. It's not just about automating infrastructure deployment; it's about creating a cohesive, well-orchestrated system where each element plays a specific role in maintaining the desired state of your infrastructure.

## Version Control and Terraform Compatibility

In Terraform Controller, the selection and management of Terraform versions are closely tied to the Runner images. Each Runner image contains a specific version of Terraform, which can be set via the chart values only, limiting the flexibility of using multiple Terraform versions simultaneously.

It's important to note that, as of now, multiple runners are not automatically released in tandem with each new version of the Terraform Controller. This issue was highlighted in a [GitHub discussion](https://github.com/weaveworks/tf-controller/issues/1023) and we're looking forward to fixing this.

> We'd also love to mention the potential support for [OpenTofu](https://opentofu.org/) in the Terraform Controller. The GitHub issue [#1132 on the Weaveworks Terraform Controller repository](https://github.com/weaveworks/tf-controller/issues/1132) responds to the growing interest in OpenTofu following HashiCorp's decision to change Terraform's licensing. OpenTofu, a fork of Terraform, remains open-source and is managed by the Linux Foundation, making it an appealing alternative for the community.

Reviewer: ```suggestion
We also love to see the upcoming potential support for [OpenTofu](https://opentofu.org/) in the Terraform Controller. The GitHub issue [#1132 on the Weaveworks Terraform Controller repository](https://github.com/weaveworks/tf-controller/issues/1132) responds to the growing interest in OpenTofu following HashiCorp's decision to change Terraform's licensing. OpenTofu, a fork of Terraform, remains open-source and is managed by the Linux Foundation, making it an appealing alternative for the community.
```

## Areas of Improvement

We've been studying the Controller for some time now, and it's clear that several important areas could be addressed to make this solution more solid.

### Implementing ChatOps for Terraform Operations

We'd like to explore the integration of ChatOps to manage Terraform operations. This includes viewing, approving, and declining pending plans and notifying about errors. Implementing ChatOps can significantly enhance communication and decision-making processes within an organization. Team members discussing infrastructure updates directly from their chat interface - wouldn't that be great?

### Utilizing Other Optional Features

We're also interested in the (Branch Planner feature)[(https://weaveworks.github.io/tf-controller/branch-planner/branch-planner-getting-started/)]. This new component detects an open PR and either creates a new Terraform object or updates an existing one, applying a Plan Only mode based on the original Terraform object. This functionality aligns really well with GitOps workflows, providing a dynamic and transparent way to manage infrastructure changes via pull requests.

### Open Source UI

> Another area of interest is the Open Source UI, as outlined in the [Weave GitOps documentation](https://docs.gitops.weave.works/docs/open-source/getting-started/ui-OSS/). The ability to easily oversee and control all the details and stages of infrastructure deployment and management is always a valuable addition.

Reviewer: ```suggestion
Another area of interest that we'd like to see built out further is the Open Source UI, as outlined in the [Weave GitOps documentation](https://docs.gitops.weave.works/docs/open-source/getting-started/ui-OSS/). The ability to easily oversee and control all the details and stages of infrastructure deployment and management is always a valuable addition.
```

## Time-sensitive: Internal Weaveworks Issues and Further Impact on the Project

While we've been working on this blog post, some quite concerning news has spread in the Weaveworks community. A recent [GitHub comment](https://github.com/weaveworks/tf-controller/issues/1166#issuecomment-1904892837) sheds some light on the fact that Weaveworks, the legal entity behind the project, is facing undisclosed issues that are impacting the direction of the project. We have to admit that the future of the Terraform Controller project seems uncertain, and the future steps of the repository's owning organization are not yet clear. It's hard to know what to expect when there is no official public statement from Weaveworks about the situation.

Despite these challenges, we still hope for the best outcome. The project is still alive, with some team members continuing their work and planning future developments. It's worth noting that the project could either be renamed and retain all repositories under the current organization or move to a different organization.

> We're happy to hear that the community is committed to continuing discussions about the project's future and development, and we're ready to support them in all possible ways.

Reviewer: ```suggestion
We're happy to hear that the community is committed to continuing discussions about the project's future and development, and we're ready to support them as believe this tool has some real potential.
```

## Conclusion

> The Terraform Controller could be a game-changer and make managing infrastructure as code in a GitOps framework a breeze. Still, we acknowledge that building a solid and usable solution requires more investments. We welcome any potential customers who are interested in learning more about this technology to reach out and chat with us!

Reviewer: ```suggestion
The Terraform Controller could be a game-changer in managing infrastructure as code via GitOps a breeze. Still, we acknowledge that building a solid and usable solution requires further investment. If you're interested in this approach, we welcome any engineering orgs who are interested in using this tool to [reach out and chat with us](https://masterpoint.io/contact/) as we'd love to build a true GitOps Terraform system for you! 
```

We would like to emphasize our positive experience with the community aspect of this product. The development team has been very responsive to our questions, providing fast and complete responses. While we understand that it may take some time to add requested features/fixes to the roadmap and implement them, we feel that the team was highly receptive to user feedback. Given recent news, we hope that everything goes as smoothly as possible for the team and the product they have been working so hard on.

Overall, we're still on a mission to explore the evolving landscape of GitOps and Terraform, and we need your help! Share your experiences and thoughts on these tools so we can work together to make them even better!

## Further Reading and Resources

> * Dive into the official [Weave GitOps Terraform Controller documentation](https://weaveworks.github.io/tf-controller/) for more profound knowledge.

Reviewer: ```suggestion
* Dive into the official [Weave GitOps Terraform Controller documentation](https://weaveworks.github.io/tf-controller/) for more in-depth knowledge.
```

Weaveworks plans to migrate the features of `tfctl` into the Weave GitOps CLI. This shift is important as the current CLI is more focused on interacting with Terraform objects rather than managing state or resources directly. Consequently, some functionalities, like the `import` command and other state-oriented operations, are currently missing. However, ongoing discussions on GitHub issues address implementing at least some of these features, indicating potential enhancements in future updates.

## Areas of Improvement

We've been studying the Controller for some time now, and it's clear that several important areas could be addressed to make this solution more solid.

### Implementing ChatOps for Terraform Operations

We'd like to explore the integration of ChatOps to manage Terraform operations. This includes viewing, approving, and declining pending plans and notifying about errors. Implementing ChatOps can significantly enhance communication and decision-making processes within an organization. Team members discussing infrastructure updates directly from their chat interface - wouldn't that be great?

### Utilizing Other Optional Features

> We're also interested in the (Branch Planner feature)[(https://weaveworks.github.io/tf-controller/branch-planner/branch-planner-getting-started/)]. This new component detects an open PR and either creates a new Terraform object or updates an existing one, applying a Plan Only mode based on the original Terraform object. This functionality aligns really well with GitOps workflows, providing a dynamic and transparent way to manage infrastructure changes via pull requests.

Author: Doh!

While we've been working on this blog post, some quite concerning news has spread in the Weaveworks community. A recent [GitHub comment](https://github.com/weaveworks/tf-controller/issues/1166#issuecomment-1904892837) sheds some light on the fact that Weaveworks, the legal entity behind the project, is facing undisclosed issues that are impacting the direction of the project. We have to admit that the future of the Terraform Controller project seems uncertain, and the future steps of the repository's owning organization are not yet clear. It's hard to know what to expect when there is no official public statement from Weaveworks about the situation.

Despite these challenges, we still hope for the best outcome. The project is still alive, with some team members continuing their work and planning future developments. It's worth noting that the project could either be renamed and retain all repositories under the current organization or move to a different organization.

We're happy to hear that the community is committed to continuing discussions about the project's future and development, and we're ready to support them as believe this tool has some real potential.

## Conclusion

> The Terraform Controller could be a game-changer in managing infrastructure as code via GitOps a breeze. Still, we acknowledge that building a solid and usable solution requires further investment. If you're interested in this approach, we welcome any engineering orgs who are interested in using this tool to [reach out and chat with us](https://masterpoint.io/contact/) as we'd love to build a true GitOps Terraform system for you!

Reviewer: ```suggestion
The Terraform Controller could be a real game-changer in managing infrastructure as code via GitOps. We believe with some further investment and improvements that it would be solid competitor to the primary open-source Terraform automation tool, [Atlantis](https://www.runatlantis.io/). Still, we acknowledge that building a strong system around this tool requires effort and our suggestion is to only entertain using this tool if your org is ready for some level of bleeding-edge investment themselves. If you're intrigued by that idea, we welcome any engineering orgs who are eager to operationalize this tool internally to [reach out and chat with us](https://masterpoint.io/contact/) as we'd love to help you build a true GitOps Terraform system!
```

We would like to emphasize our positive experience with the community aspect of this product. The development team has been very responsive to our questions, providing fast and complete responses. While we understand that it may take some time to add requested features/fixes to the roadmap and implement them, we feel that the team was highly receptive to user feedback. Given recent news, we hope that everything goes as smoothly as possible for the team and the product they have been working so hard on.

> Overall, we're still on a mission to explore the evolving landscape of GitOps and Terraform, and we need your help! Share your experiences and thoughts on these tools so we can work together to make them even better!

Reviewer: ```suggestion
Overall, we're still on a mission to explore the evolving landscape of GitOps and Terraform, and we need your help! [Get in touch with us on LinkedIn](https://www.linkedin.com/company/masterpoint-consulting/) so you can share your experiences + thoughts so we can work together to make them even better!
```

## Further Reading and Resources

> * Dive into the official [Weave GitOps Terraform Controller documentation](https://weaveworks.github.io/tf-controller/) for more in-depth knowledge.

Reviewer: ```suggestion
* Dive into the official [Weaveworks GitOps Terraform Controller documentation](https://weaveworks.github.io/tf-controller/) for more in-depth knowledge.
```
